### DISCUSSION TITLE
Monocular Vision

### DISCUSSION THREAD

**Discussion by PranavU-Coder** [MITB-AUVTeam/Final_AUV#10] [Ideas]
# Monocular Vision -> for Depth Estimation and more such tasks

Depth estimation is a vital information for our torpedo launcher, how are you supposed to know how far something has to be shot especially when the target's diameter is barely 3 cm ?

![shark_on_top](https://github.com/user-attachments/assets/2722a9c1-7903-4de3-a77e-6bd297e63511)

_talking wrt the hole in the displayed image_

In such cases , precision matters more than anything and so all ml based depth models were under covered for this.

why ml-based depth models ? you can't really get the actual estimate of how far something is using a monocular camera cause there is no "parallax" or disparity

![stereovision](https://github.com/user-attachments/assets/34bd0eac-357d-4db2-a1c1-7cccf8c1a193)

_disparity = x - x'  = Bf/Z_

so x and x' are the distance between points in image plane corresponding to the scene point 3D and their camera center. is the distance between two cameras (which we know) and is the focal length of camera (already known). So in short, the above equation says that the depth of a point in a scene is inversely proportional to the difference in distance of corresponding image points and their camera centers. So with this information, we can derive the depth of all pixels in an image.

unfortunately for that you would need two cameras and as such disparity can't be obtained and thus no estimation or calculation.

this is where ML-depth models actually come into use , however it is important that all ML-models can never estimate but instead can create a prediction that can be near accurate as they make guesses based on what they have been trained prior.

## ML Models Utilized For Monocular Depth Estimation

- ml-depth-pro
- DepthAnythingV2
- MiDaS
- ZoeDepth

before experimenting with these high level models , I tried creating my own depth maps and tried to estimate from those , however those attempts didn't bear any fruits.

## ml-depth-pro

apple's ml-depth-pro was definitely the most promising depth estimation model cause it does actually estimate the depth of what you're looking for if you can couple it with your YOLO model (object detection model for real time use)
``` python
import cv2 as cv
import numpy as np

from ultralytics import YOLO
import depth_pro

model = YOLO('/home/pranav/Projects/Final_AUV/torpedo_launcher/bounding_boxes/runs/detect/train/weights/best.pt')

img_path = 'image.jpg'
img_input = cv.imread(img_path)

if img_input is None:
    raise FileNotFoundError(f"Unable to load image at {img_path}")

results = model(img_input)

target_boxes = []
target_classes = ['hole']

for result in results:
    boxes = result.boxes.xyxy.cpu().numpy()
    classes = result.boxes.cls.cpu().numpy()

    for box, cls in zip(boxes, classes):
        name = result.names[int(cls)]
        if name in target_classes:
            x1, y1, x2, y2 = map(int, box[:4])
            target_boxes.append((x1, y1, x2, y2, name))
            cv.rectangle(img_input, (x1, y1), (x2, y2), (0, 255, 0), 2)

depth_model, transform = depth_pro.create_model_and_transforms()
depth_model.eval()

img, _, f_px = depth_pro.load_rgb(img_path)
img = transform(img).unsqueeze(0)

predict = depth_model.infer(img, f_px=f_px)
depth = predict['depth']

depth_np = depth.squeeze().cpu().numpy()

depth_min, depth_max = depth_np.min(), depth_np.max()
range_depth = depth_max - depth_min if depth_max != depth_min else 1.0

for x1, y1, x2, y2, name in target_boxes:
    center_x = np.clip((x1 + x2) // 2, 0, depth_np.shape[1] -1)
    center_y = np.clip((y1 + y2) // 2, 0, depth_np.shape[0] -1)
    depth_value = depth_np[center_y, center_x]
    text = f'{name} depth: {depth_value:.2f}m'

    font = cv.FONT_HERSHEY_SIMPLEX
    font_scale = 1
    font_thickness = 2
    text_size = cv.getTextSize(text, font, font_scale, font_thickness)[0]

    text_x = x1
    text_y = y1 - 10

    rect_x1 = text_x - 5
    rect_y1 = text_y - text_size[1] - 10
    rect_x2 = text_x + text_size[0] + 5
    rect_y2 = text_y + 5

    cv.rectangle(img_input, (rect_x1, rect_y1), (rect_x2, rect_y2), (0, 0, 0), -1)
    cv.putText(img_input, text, (text_x, text_y), font, font_scale, (255, 255, 255), font_thickness)

cv.namedWindow('Annotated Image', cv.WINDOW_NORMAL)
cv.imshow('Annotated Image', img_input)
cv.imwrite('depth.jpg', img_input)
cv.waitKey(0)

depth_np_normalized = (depth_np - depth_min) / range_depth
inv_depth_np_normalized = 1.0 - depth_np_normalized
depth_colormap = cv.applyColorMap((inv_depth_np_normalized * 255).astype(np.uint8), cv.COLORMAP_TURBO)

cv.namedWindow('Depth Colormap', cv.WINDOW_NORMAL)
cv.imshow('Depth Colormap', depth_colormap)
cv.imwrite('depth_colormap.jpg', depth_colormap)
cv.waitKey(0)

cv.destroyAllWindows()
```

_the code I wrote for coupling yolo model with ml_depth_pro which it took me around 2 days_ 

however there is one small issue -> it takes about 3 minutes to process and give you back the result

which means it is not really feasible for real time shooting scenario , which is a shame since ml-depth-pro is the one that actually gave a pretty accurate result compared to all of the other depth estimated models I tinkered with.

_important to note these models don't return the vertical or horizontal distance value but rather the 'hypotenuse' (slant height if you can call that)_

 
## DepthAnythingV2

now regular DepthAnythingV2's models are just used for creating depth maps , after digging deeper I found metric-estimation models that they provide as well.

DepthAnythingV2's metric depth estimation models come in two flavors -> hypersim (for indoor purposes) and vkitti (for outdoor purposes) and three more versions based on parameter count so in total 6 models.

I will be honest , the results were underwhelming for all 6 models and I was heavily disappointed that none of the models were able to get the job done.

![depth_colormap_vitl](https://github.com/user-attachments/assets/2b14d07a-4755-4568-9259-c268ac8cb3c3)

_an example of the depth colormap that DepthAnythingV2's large vision transformer created_

## MiDaS

MiDaS is a depth estimation model similar to DepthAnythingV2 however again it is effective just to create depth maps and thus the experimentation with MiDaS was short-lived.

## ZoeDepth

a _zero_ _shot_ depth estimation model , which unfortunately is no longer being improved on and has been archived , I couldn't properly utilize it and I can only assume how the results would be given how zero-shot models I have worked with in past has given me.

# Conclusion

Now that we are going to use a dedicated depth camera in future , I am no longer going to experiment with these models , however as mentioned it can be a great alternative if a dedicated stereocameras or depth cameras is not affordable also apple's ml-depth-pro results were actually good and would be a genuine option had it not taken around 3 minutes to calculate and return results.

# References 

- [how to create depth maps and functioning of stereocameras ](https://docs.opencv.org/4.x/dd/d53/tutorial_py_depthmap.html)
- [robosub team that uses ML-models for tasks](https://robonation.org/app/uploads/sites/4/2024/07/RS24_TDR_IstanbulTechU.pdf)
- [apple's ml-depth-pro model](https://github.com/apple/ml-depth-pro)
- [DepthAnythingV2](https://github.com/DepthAnything/Depth-Anything-V2/tree/main/metric_depth)
- [MiDaS](https://github.com/isl-org/MiDaS)
- [ZoeDepth](https://github.com/isl-org/ZoeDepth)
 
in reference to #6

**Comment by Probatio-Diabolica** [comment]
Great job! so what's your next experiment plan?

**Reply by PranavU-Coder** [reply]
well no more experimentation , now lockin in with stereocameras :fire:
